### DISCUSSION TITLE
Putting dom reference into store will cause lag

### DISCUSSION THREAD

**Discussion by MilesLeeeeee** [pmndrs/zustand#2079] [Bug report] [answered]
## Summary
Hello
I tried putting the dom reference into the store to make it easier for me to use, but I found that it made the browser lag.

I tested in three browsers

chrome：117
edge：117
firefox：116

Chrome and edge are experiencing serious lags

firefox runs fine

Is there something wrong with my operation or should I not put the dom reference into the store?

## Link to reproduction
[https://codesandbox.io/s/zustand-dom-test-d7g4l3?file=/src/App.js](https://codesandbox.io/s/zustand-dom-test-d7g4l3?file=/src/App.js)

You can see the difference by changing the comment on this line

![image](https://github.com/pmndrs/zustand/assets/129132865/2262e76c-a3e2-4af0-9cc4-52c097e293a9)

**Comment by dai-shi** [comment]
Can anyone look into this?

**Comment by Pong420** [comment, marked as answer]
It should because of the  `devtools` middleware. Remove it or adding the `stateSanitizer` option.
```ts
{
  name: "bear",
  stateSanitizer: (state) => {
    return JSON.parse(
      JSON.stringify(state, (_key, value) => {
        if (typeof value === "function") return undefined;
        if (value instanceof HTMLElement)
          return `${value.tagName} ${value.className}`;
        if (value instanceof Date) return value.toLocaleString();
        return value;
      })
    );
  },
}
```

<img width="2405" height="1455" alt="Screenshot 2026-01-03 at 7 43 03 PM" src="https://github.com/user-attachments/assets/e633b300-b74f-477c-b44d-ce5033fe4333" />

The root cause should be related to  https://github.com/facebook/react/issues/24360
setting this property also work
```ts
HTMLElement.prototype.toJSON = () => "{}";
```

**Reply by MilesLeeeeee** [reply]
Yes, I tried the two methods you mentioned, and they did solve the problem. I didn't realize the problem was so deeply hidden. Thank you for your answer.

**Comment by vincedsb1** [comment]
To avoid browser lag when putting DOM references into a Zustand store, you can utilize the `stateSanitizer` option or remove the devtools middleware. This can help prevent performance issues, especially in Chrome and Edge browsers.

Here's an example of how you can configure your store to use the `stateSanitizer` option:

```jsx
{
  name: "bear",
  stateSanitizer: (state) => {
    return JSON.parse(
      JSON.stringify(state, (_key, value) => {
        if (typeof value === "function") return undefined;
        if (value instanceof HTMLElement)
          return `${value.tagName} ${value.className}`;
        if (value instanceof Date) return value.toLocaleString();
        return value;
      })
    );
  },
}
```

This configuration helps sanitize the state, preventing potential performance issues caused by DOM references in Chrome and Edge browsers.

Additionally, you can set `HTMLElement.prototype.toJSON` to an empty object to address the root cause related to the issue mentioned in the Facebook React repository.

Hope this helps! If it solves your issue, could you please mark this comment as the answer? It helps others find the solution faster. 🙏

**Reply by MilesLeeeeee** [reply]
Yes, both methods can solve the problem. Thank you for your answer. 👍
